pkg/kvstore/etcd_debug.go

```diff
@@ -0,0 +1,398 @@
+			etcdDbgOutputCert(leaf, iiw)
+			iiw = iiw.WithExtraIndent(2)
+
+			// Print intermediate certificates, if any.
+			intermediates := x509.NewCertPool()
+			for _, cert := range cert.Certificate[1:] {
+				iiw.Println("Intermediates:")
+
+				intermediate, err := x509.ParseCertificate(cert)
+				if err != nil {
+					iw.Println("❌ Failed to parse intermediate certificate: %s", err)
+					continue
+				}
+
+				etcdDbgOutputCert(intermediate, iiw)
+				intermediates.AddCert(intermediate)
+			}
+
+			// Attempt to verify whether the given certificate can be validated
+			// using the configured root CAs. Although a failure is not necessarily
+			// an error, as the remote etcd server may be configured with a different
+			// root CA, it still signals a misconfiguration in most cases.
+			opts := x509.VerifyOptions{
+				Roots:         cfg.TLS.RootCAs,
+				Intermediates: intermediates,
+			}
+
+			_, err = leaf.Verify(opts)
+			if err != nil {
+				iiw.Println("⚠️ Cannot verify certificate with the configured root CAs")
+			}
+		}
+	}
+
+	if cfg.Username != "" {
+		passwd := "unset"
+		if cfg.Password != "" {
+			passwd = "set"
+		}
+
+		iw.Println("✅ Username set to %s, password is %s", cfg.Username, passwd)
+	}
+}
+
+func etcdDbgOutputIPs(ips []net.IP) string {
+	var buf bytes.Buffer
+	for i, ip := range ips {
+		if i > 0 {
+			buf.WriteString(", ")
+		}
+
+		if i == 4 {
+			buf.WriteString("...")
+			break
+		}
+
+		buf.WriteString(ip.String())
+	}
+	return buf.String()
+}
+
+func etcdDbgRetrieveRootCAFile(cfgfile string) (certs [][]byte, err error) {
+	var yc yamlConfig
+
+	b, err := os.ReadFile(cfgfile)
+	if err != nil {
+		return nil, err
+	}
+
+	err = yaml.Unmarshal(b, &yc)
+	if err != nil {
+		return nil, err
+	}
+
+	crtfile := cmp.Or(yc.TrustedCAfile, yc.CAfile)
+	if crtfile == "" {
+		return nil, errors.New("not provided")
+	}
+
+	data, err := os.ReadFile(crtfile)
+	if err != nil {
+		return nil, err
+	}
+
+	for {
+		block, rest := pem.Decode(data)
+		if block == nil {
+			if len(certs) == 0 {
+				return nil, errors.New("no certificate found")
+			}
+
+			return certs, nil
+		}
+
+		if block.Type == "CERTIFICATE" {
+			certs = append(certs, block.Bytes)
+		}
+
+		data = rest
+	}
+}
+
+func etcdDbgOutputCert(cert *x509.Certificate, iw *indentedWriter) {
+	sn := cert.SerialNumber.Text(16)
+	for i := 2; i < len(sn); i += 3 {
+		sn = sn[:i] + ":" + sn[i:]
+	}
+
+	iw.Println("- Serial number:       %s", string(sn))
+	iw.Println("  Subject:             %s", cert.Subject)
+	iw.Println("  Issuer:              %s", cert.Issuer)
+	iw.Println("  Validity:")
+	iw.Println("    Not before:  %s", cert.NotBefore)
+	iw.Println("    Not after:   %s", cert.NotAfter)
+}
+
+type indentedWriter struct {
+	w      io.Writer
+	indent []byte
+}
+
+func newIndentedWriter(w io.Writer, indent int) *indentedWriter {
+	return &indentedWriter{w: w, indent: []byte(strings.Repeat(" ", indent))}
+}
+
+func (iw *indentedWriter) NewLine() { iw.w.Write([]byte("\n")) }
```

Comment:
nit: Is there a reason for these to be exported?

Reply:
Not really. The idea was just to explicit that these are the methods expected to be used by the consumers of `indentedWriter` (although there aren't others at the moment).